src/scene.rs

```diff
@@ -1,25 +1,95 @@
-use bevy::{prelude::*, scene::SceneInstance};
+use bevy::{
+    prelude::*,
+    scene::{SceneInstance, SceneInstanceReady},
+};
 
-use crate::InheritOutlineBundle;
+use crate::{compute_outline, InheritOutlineBundle};
 
 /// A component for triggering the `AsyncSceneInheritOutlinePlugin`.
 #[derive(Component)]
 pub struct AsyncSceneInheritOutline;
 
-fn process_async_scene_outline(
+/// A component marking that `AsyncSceneInheritOutlinePlugin` has processed a scene.
+#[derive(Component)]
+pub struct AsyncSceneInheritOutlineProcessed;
+
+fn instance_maybe_ready(
+    mut commands: Commands,
+    scene_spawner: &SceneSpawner,
+    entity: Entity,
+    instance: &SceneInstance,
+) {
+    if scene_spawner.instance_is_ready(**instance) {
+        for child in scene_spawner.iter_instance_entities(**instance) {
+            commands
+                .entity(child)
+                .insert(InheritOutlineBundle::default());
+        }
+        commands
+            .entity(entity)
+            .insert(AsyncSceneInheritOutlineProcessed);
+    }
+}
+
+// Handles scenes which are already ready when `AsyncSceneInheritOutline` is added.
+#[allow(clippy::type_complexity)]
+fn async_added(
+    mut commands: Commands,
+    scene_spawner: Res<SceneSpawner>,
+    async_query: Query<
+        (Entity, &SceneInstance),
+        (
+            Added<AsyncSceneInheritOutline>,
+            Without<AsyncSceneInheritOutlineProcessed>,
+        ),
+    >,
+) {
+    for (entity, instance) in async_query.iter() {
+        instance_maybe_ready(commands.reborrow(), &scene_spawner, entity, instance);
+    }
+}
+
+// Handles scenes which become ready after `AsyncSceneInheritOutline` is added.
+#[allow(clippy::type_complexity)]
+fn async_ready_event(
```

Comment:
Why do you need both systems? Isn't `SceneInstanceReady` emitted every time after an instance spawn?

Reply:
This supports the use case that you first spawn a scene, the scene becomes ready, and then only later add the outlining components to it.

Personally, I always add the outline components at spawning time and never remove them, but I can't say that an approach where you add and remove the components is wrong. I don't know how efficient the `Added` query filter is though, so perhaps this is no better than not using the event.

Reply:
Makes sense.

`Added` query is a syntax sugar for iterating and checking `is_added()` on `Ref` or `&mut` components.